File: ruby/crypto-square/crypto_square.rb
class Crypto
  def initialize(text)
    @source_text = normalise(text)
  end

  def ciphertext
    return source_text if source_text.empty?
    encode
  end

  private

  attr_reader :source_text

  def normalise(text)
    text.downcase.scan(/[a-z1-9]/).join
  end

  def encode
    generate_rectangle(source_text).transpose.map(&:join).join(' ')
  end

  def generate_rectangle(string)
    apply_padding(string).chars.each_slice(columns).to_a
  end

  def apply_padding(string)
    padding = (columns * rows) - string.size
    string.concat(' ' * padding)
  end

  def columns
    n = Math.sqrt(source_text.size).ceil
    n.zero? ? 1 : n
  end

  def rows
    (source_text.size / columns.to_f).ceil
  end
end

module BookKeeping
  VERSION = 1
end

File: python/crypto-square/crypto_square.py
import math

def encode(plain_text):
    return ' '.join([''.join(t) for t in list(zip(*generate_rectangle(normalise(plain_text))))])

def generate_rectangle(normalised_text):
    text = apply_padding(normalised_text)
    return [text[i:i + columns(text)] for i in range(0, len(text), columns(text))]

def apply_padding(text):
    padding_value = (columns(text) * rows(text)) - len(text)
    return text + ' ' * padding_value

def normalise(text):
    return str.lower(''.join([char for char in text if char.isalnum()]))

def columns(text):
    return math.ceil(math.sqrt(len(text))) or 1

def rows(text):
    return math.ceil(len(text) / columns(text))
